tests/zfs-tests/include/libtest.shlib

```diff
@@ -1449,23 +1449,113 @@ function reexport_pool
 }
 
 #
-# Verify a given disk is online or offline
+# Verify a given disk or pool state
 #
 # Return 0 is pool/disk matches expected state, 1 otherwise
 #
-function check_state # pool disk state{online,offline}
+function check_state # pool disk state{online,offline,degraded}
 {
 	typeset pool=$1
 	typeset disk=${2#$DEV_DSKDIR/}
 	typeset state=$3
 
-	$ZPOOL status -v $pool | grep "$disk"  \
-	    | grep -i "$state" > /dev/null 2>&1
+	[[ -z $pool ]] || [[ -z $state ]] \
+	    && log_fail "Arguments invalid or missing"
+
+	if [[ -z $disk ]]; then
+		#check pool state only
+		$ZPOOL get -H -o value health $pool \
+		    | grep -i "$state" > /dev/null 2>&1
+	else
+		$ZPOOL status -v $pool | grep "$disk"  \
+		    | grep -i "$state" > /dev/null 2>&1
+	fi
 
 	return $?
 }
 
 #
+# Online or offline a disk on the system
+#
+# First checks state of disk. Test will fail if disk is not properly onlined
+# or offlined
+#
+function on_off_disk # disk state{online,offline} HBA CHAN TARG LUN
+{
+	typeset disk=$1
+	typeset state=$2
+	typeset HBA=$3
+	typeset CHAN=$4
+	typeset TARG=$5
+	typeset LUN=$6
+
+	[[ -z $disk ]] || [[ -z $state ]] || [[ -z $HBA ]] || \
+	    [[ -z $CHAN ]] || [[ -z $TARG ]] || [[ -z $LUN ]] \
+	    && log_fail "Arguments invalid or missing"
+
+	if is_linux; then
+		if [[ $state == "offline" ]] && ( is_mpath_device $disk ); then
+			dm_name="$($READLINK $DEV_DSKDIR/$disk \
+			    | $NAWK -F / '{print $2}')"
+			slave="$($LS /sys/block/$dm_name/slaves \
+			    | $NAWK '{print $1}')"
+			#check if disk is online
+			$LSSCSI | $EGREP $slave > /dev/null
+			if (($? == 0)); then
+				$ECHO "offline" > \
+				    /sys/class/scsi_device/${HBA}:${CHAN}:${TARG}:${LUN}/device/state
+				$ECHO "1" > \
+				    /sys/class/scsi_device/${HBA}:${CHAN}:${TARG}:${LUN}/device/delete
+				$LSSCSI | $EGREP $slave > /dev/null
+					if (($? == 0)); then
+						log_fail "Offlining $disk failed"
+					fi
+			else
+				log_note "$disk is already offline"
+			fi
+		elif [[ $state == "offline" ]] && ( is_real_device $disk ); then
+			#check if disk is online
+			$LSSCSI | $EGREP $disk > /dev/null
+			if (($? == 0)); then
+				$ECHO "offline" > \
+				    /sys/block/$disk/device/state
+				$ECHO "1" > \
+				    /sys/block/$disk/device/delete
+				$LSSCSI | $EGREP $disk > /dev/null
+					if (($? == 0)); then
+						log_fail "Offlining $disk failed"
+					fi
+			else
+				log_note "$disk is already offline"
+			fi
+		elif [[ $state == "online" ]]; then
+			$ECHO "$CHAN $TARG $LUN" > \
+			    /sys/class/scsi_host/host${HBA}/scan
```

Comment:
I suspect forcing a full rescan would be a simpler and more robust way to handle this.  Plus it's the kind of thing people will do and it's need to not cause problems.  As you pointed out RedHat provides a `scsi-rescan` tool as do other distributions.